Add runwoke.sh for "tox -e woke" - Concious language

nhosoi opened the pull request:
This approach might be too naive to use the `woke` tool, but we could run the test locally by `tox -e woke` ~~and also via CI~~.

The script runwoke.sh installs the latest `woke` from `https://git.io/getwoke`, copies woke.yml from the config_files dir to the top level of the role to be checked, then runs `woke` using the config file. The output from woke is found in .tox/woke/tmp/woke_output.txt.

The current generated `.woke.yml` looks like this. I added `master` and `slave` to adjust to the suggested words by Red Hat.
```
ignore_files:
  - .tox
  - .github
  - .sanity-ansible-ignore-*.txt
  - pylintrc

rules:
  - name: master
    terms:
      - master
    alternatives:
      - primary

  - name: slave
    terms:
      - slave
    alternatives:
      - secondary

  - name: sanity
    terms:
      - sanity
    alternatives:
      - confidence
      - quick check
      - coherence check
    severity: warning

  - name: dummy
    terms:
      - dummy
    alternatives:
      - placeholder
      - sample
    severity: warning
```
Some questions came up in the testing...
1) For instance, to skip this error
```
CHANGELOG.md:179:57-63: `sanity` may be insensitive, use `confidence`, `quick check`, `coherence check` instead (error)
- update to tox-lsr 2.4.0 - add support for ansible-test sanity with docker
```
Is adding `// wokeignore:rule=sanity` at the end of the line acceptable?
```
-- update to tox-lsr 2.4.0 - add support for ansible-test sanity with docker
+- update to tox-lsr 2.4.0 - add support for ansible-test sanity with docker // wokeignore:rule=sanity
 - use tox-ls
```
The word `sanity` is treated as a warning.

2) There are quite a lot of words used in the products themselves (nbde_client).
E.g., a function name in the library.
```
library/nbde_client_clevis.py:1435:13-19: `sanity` may be insensitive, use `confidence`, `quick check`, `coherence check` instead (error)
def bindings_sanity_check(bindings, data_dir, check_mode):
             ^
```
Another example of using `Master` is in the config param (ssh).
```
tests/tests_global_config.yml:21:17-23: `Master` may be insensitive, use `primary` instead (error)
          ControlMaster: auto
                 ^
```
What we should do to workaround these errors? Shall we set `severity: warning` to `master` and `sanity` and leave them as they are?

The word `sanity` is treated as a warning.

richm commented on src/tox_lsr/test_scripts/runwoke.sh:
```diff
@@ -0,0 +1,56 @@
+
+if grep 'instead (error)' "$WOKE_OUT" 2>&1 /dev/null ; then
+  cat "$WOKE_OUT"
+  lsr_error "${ME}: Concious language violation was reported."
```
```suggestion
  lsr_error "${ME}: Conscious language violation was reported."
```

nhosoi commented:
Thank you, @richm. I fixed the typo and added `severity: warning` to the rule for `sanity`.

> For github CI, we'll use https://github.com/marketplace/actions/run-woke instead of tox-lsr

BTW, what would be a good way to propagate the .woke.yml file to the action? Can we ask each role to push the generated .woke.yml to the repo? Or could there be any good place to put to download it in the action, or ...?

richm commented:
> Thank you, @richm. I fixed the typo and added `severity: warning` to the rule for `sanity`.
> 
> > For github CI, we'll use https://github.com/marketplace/actions/run-woke instead of tox-lsr
> 
> BTW, what would be a good way to propagate the .woke.yml file to the action? Can we ask each role to push the generated .woke.yml to the repo? Or could there be any good place to put to download it in the action, or ...?

The underlying `woke` tool supports loading a config file from a url: https://github.com/get-woke/woke/blob/main/docs/usage.md#remote-config-file
I think we should keep the config file in tox-lsr
The woke action has a `woke-args` parameter.  maybe we could do something like this for the action:
```
      - name: woke
        uses: get-woke/woke-action@v0
        with:
          woke-args: "-c https://raw.githubusercontent.com/linux-system-roles/tox-lsr/main/src/tox_lsr/config_files/woke.yml"
```
or something like that

richm commented:
lgtm - you'll need to update the result.ini test file - do
```
UPDATE_RESULT_INI=true tox -e py310-tox30
```
then add the updated file to the commit

nhosoi commented:
> lgtm - you'll need to update the result.ini test file - do
> 
> ```
> UPDATE_RESULT_INI=true tox -e py310-tox30
> ```
> 
> then add the updated file to the commit

Thanks! It added this config. (Not yet committed, though.)
```
diff --git a/tests/fixtures/test_tox_merge_ini/result.ini b/tests/fixtures/test_tox_merge_ini/result.ini
index 86ed245..9406711 100644
--- a/tests/fixtures/test_tox_merge_ini/result.ini
+++ b/tests/fixtures/test_tox_merge_ini/result.ini
@@ -224,6 +224,10 @@ basepython = python3
 deps = {env:LSR_ANSIBLE_TEST_DEP:ansible-core==2.14.*}
 commands = bash {lsr_scriptdir}/runansible-test.sh
 
+[testenv:woke]
+changedir = {toxinidir}
+commands = bash {lsr_scriptdir}/runwoke.sh
+
 [qemu_common]
 changedir = {toxinidir}
 basepython = python3
```
Does this mean, for instance, this tox test would include `woke` run?
tox / python (3.10, ubuntu-latest) (pull_request) Successful in 40s
[Details](https://github.com/linux-system-roles/logging/actions/runs/3650221768/jobs/6165933155)
If so, will it be running `woke` in the tox test as well as in the action?

nhosoi commented:
@richm, ^^^^^?

richm commented:
> > lgtm - you'll need to update the result.ini test file - do
> > ```
> > UPDATE_RESULT_INI=true tox -e py310-tox30
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then add the updated file to the commit
> 
> Thanks! It added this config. (Not yet committed, though.)
> 
> ```
> diff --git a/tests/fixtures/test_tox_merge_ini/result.ini b/tests/fixtures/test_tox_merge_ini/result.ini
> index 86ed245..9406711 100644
> --- a/tests/fixtures/test_tox_merge_ini/result.ini
> +++ b/tests/fixtures/test_tox_merge_ini/result.ini
> @@ -224,6 +224,10 @@ basepython = python3
>  deps = {env:LSR_ANSIBLE_TEST_DEP:ansible-core==2.14.*}
>  commands = bash {lsr_scriptdir}/runansible-test.sh
>  
> +[testenv:woke]
> +changedir = {toxinidir}
> +commands = bash {lsr_scriptdir}/runwoke.sh
> +
>  [qemu_common]
>  changedir = {toxinidir}
>  basepython = python3
> ```
> 
> Does this mean, for instance, this tox test would include `woke` run? tox / python (3.10, ubuntu-latest) (pull_request) Successful in 40s [Details](https://github.com/linux-system-roles/logging/actions/runs/3650221768/jobs/6165933155)

No.  If you'll notice https://github.com/linux-system-roles/logging/actions/runs/3650221768/workflow#L56 the test environments to run are explicitly specified - so you would have to add `woke` there.  But we don't want to do that - we want to have a completely separate github action (.github/workflows/woke.yml) file

> If so, will it be running `woke` in the tox test as well as in the action?

nhosoi commented:
Thank you for the confirmation, @richm. I've updated this PR as you suggested and all the CI tests have passed!

richm commented on src/tox_lsr/test_scripts/runwoke.sh:
```diff
@@ -0,0 +1,29 @@
+
+trap "rm -f $TOXINIDIR/.woke.yml" RETURN
+
+pushd "$TOXINIDIR" > /dev/null
```
You shouldn't have to pushd or cd to `$TOXINIDIR` because the testenv:woke does `changedir = {toxinidir}`

richm commented on src/tox_lsr/test_scripts/runwoke.sh:
```diff
@@ -0,0 +1,29 @@
+trap "rm -f $TOXINIDIR/.woke.yml" RETURN
+
+pushd "$TOXINIDIR" > /dev/null
+cp "$LSR_CONFIGDIR/woke.yml" .woke.yml
```
why copy?

richm commented on src/tox_lsr/test_scripts/runwoke.sh:
```diff
@@ -0,0 +1,29 @@
+
+pushd "$TOXINIDIR" > /dev/null
+cp "$LSR_CONFIGDIR/woke.yml" .woke.yml
+"$WOKE_CMD" > "$WOKE_OUT" 2>&1
```
can't you use `"$WOKE_CMD" -c "$LSR_CONFIGDIR/woke.yml"` ?

richm replied:
Also, with the `set -e` above - if `"$WOKE_CMD"` exits with a non-zero exit code, won't the script exit there?  I think you'll need to use `"$WOKE_CMD" -c "$LSR_CONFIGDIR/woke.yml" > "$WOKE_OUT" 2>&1 || :` to ignore any error return from woke

nhosoi commented:
Thank you, @richm. I've applied the changes you suggested.

nhosoi commented:
@richm, @spetrosi, if this PR looks ok, could you please merge it? Once it's done, the action in each role can refer to the customized config added in this PR. Thanks!
